src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -3523,19 +3544,136 @@ WorkerSortClauseList(MultiExtendedOp *originalOpNode,
 
 		if (orderByNonAggregates)
 		{
-			workerSortClauseList = list_copy(sortClauseList);
+			workerSortClauseList = sortClauseList;
 			workerSortClauseList = list_concat(workerSortClauseList, groupClauseList);
 		}
 		else if (canApproximate)
 		{
-			workerSortClauseList = originalOpNode->sortClauseList;
+			workerSortClauseList = sortClauseList;
 		}
 	}
 
 	return workerSortClauseList;
 }
 
 
+/*
+ * AdjustWorkerSortClausesAndTargetList goes over provided sort clause lists and
+ * make sure they still reference to a correct target entry.
+ *
+ * The function is required because we change target entry if it contains an
+ * aggregate operation, AVG operation specific to this case. A problem arises
+ * if there is an order by on an average expression. Worker query is changed
+ * to have 2 target entries, sum() and count(), if we have order by and limit
+ * on this target entry we want to (1) use correct target entry (avg), and (2)
+ * point to this particular target entry from sort clauses.
+ *
+ * Note the function modifies workerExtendedOpNodes targetList, contents
+ * of SortGroupClause items in sortClauseList, and nextSortGroupRefIndex.
+ */
+static void
+AdjustWorkerSortClausesAndTargetList(MultiExtendedOp *workerExtendedOpNode,
+									 MultiExtendedOp *originalOpNode,
```

Comment:
Could we pass `originalOpNode -> targetList ` instead of passing all?

> + * Note the function modifies workerExtendedOpNodes targetList, contents
+ * of SortGroupClause items in sortClauseList, and nextSortGroupRefIndex.

Hmm, also should we pass `workerExtendedOpNode->targetList` and return the modified list? I think it's acceptable to modify `nextSortGroupRefIndex` and elements of `sortClauseList `. But, I feel like it'd be more readable if we return the new targetlist from the function.

Reply:
we are modifying all passed arguments. Selectively returning one of them looked weird to me.

Reply:
But, modifying 3 parameters in a single function seems weird as well. I don't think we have any other code that does it. 

Even if we have some functions which *mutates* the input, their name included *mutate*.

And, if we keep the API like this, how are we going to use it with window functions?

Reply:
I propose the following API, which seems to be easliy applicable to window functions as well:

```C
static List *
AddRequiredSortExpressionsToTargetList(List *queryTargetList, List *sortGroupClauseList, Index *nextSortGroupRefIndex)
{
    /* 
      * modify elementsOf sortList as needed, 
       * increment nextSortGroupRefIndex as needed, 
      *return new targetList*/
     return newTargetListWithNewElements
}
```

Here:

```C
workerExtendedOpNode->targetList = AddRequiredSortExpressionsToTargetList (workerExtendedOpNode->targetList, sortClauseList, &refIndex)
```


And on window functions:
```C
targetList = AddRequiredSortExpressionsToTargetList(targetList, window->partitionClause, &refIndex);
targetList = AddRequiredSortExpressionsToTargetList(targetList, window->orderClause, &refIndex);
```

Reply:
We need to access target list of original extended op node since we are modifying the worker target list up to this point.

Reply:
per our discussion
1 - function will generate and return list of new target entries to be appended to worker target list
2 - it will be possible to have repetitions of the same aggregate if it is also used in window function order by clause